Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+theorem reverse_accepts {M : NFA α σ} :
+    M.reverse.accepts = { xs : List α | xs.reverse ∈ M.accepts } := by
+  ext xs
+  rw [accepts_acceptsFrom, accepts_acceptsFrom, reverse_acceptsFrom_rewindsToStart, Set.mem_setOf]
+  apply reverse_rewindsToStart_acceptsFrom
+
+end Reversal
+
+section Union
+
+variable {σ1 : Type v} {σ2 : Type v}
+
+private instance Language.instUnion : Union (Language α) := by
+  apply Set.instUnion
+
+/--`stepSum M₁ M₂` computes the transition for `M₁ ∪ M₂`. -/
+def stepSum (M1 : NFA α σ1) (M2 : NFA α σ2)
+  (s : σ1 ⊕ σ2) (a : α) : Set (σ1 ⊕ σ2) :=
+    { s' : σ1 ⊕ σ2 |
+      Sum.LiftRel
+        (fun s1 s1' ↦ s1' ∈ M1.step s1 a)
+        (fun s2 s2' ↦ s2' ∈ M2.step s2 a)
+        s s' }
+
+/-- NFAs are closed under union:
+  Given NFAs `M₁` and `M₂`, `M₁ ∪ M₂` is a NFA such that
+  `L(M₁ ∪ M₂) = L(M₁) ∪ L(M₂)`. -/
+def union (M1 : NFA α σ1) (M2 : NFA α σ2) : NFA α (σ1 ⊕ σ2) where
+  start : Set (σ1 ⊕ σ2) := { s : σ1 ⊕ σ2 | s.casesOn M1.start M2.start }
+  step : σ1 ⊕ σ2 → α → Set (σ1 ⊕ σ2) := stepSum M1 M2
+  accept : Set (σ1 ⊕ σ2) := { s : σ1 ⊕ σ2 | s.casesOn M1.accept M2.accept }
+
+lemma union_biUnion_spec
+  {x : α} {S1 : Set σ1} {S2 : Set σ2} {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    (⋃ s, ⋃ (_ : Sum.rec S1 S2 s), stepSum M1 M2 s x)
+    =
+    {s' | Sum.rec
+      (fun s1' ↦ (⋃ (s1 : σ1) (_ : s1 ∈ S1), M1.step s1 x) s1')
+      (fun s2' ↦ (⋃ (s2 : σ2) (_ : s2 ∈ S2), M2.step s2 x) s2') s'} := by
+  ext s'
+  rw [Set.mem_iUnion, Set.mem_setOf]
+  dsimp [stepSum]
+  cases s'
+  case inl s1' =>
+    simp
+    rw [←Set.mem_def (a := s1') (s := (⋃ (s1 : σ1) (_ : s1 ∈ S1), M1.step s1 x))]
+    rw [Set.mem_iUnion]
+    constructor
+    · rintro ⟨s1, hs1, h1⟩
+      exists s1
+      rw [Set.mem_iUnion]
+      exists hs1
+    · rintro ⟨s1, hs1⟩
+      rw [Set.mem_iUnion] at hs1
+      rcases hs1 with ⟨hs1,h1⟩
+      exists s1
+  case inr s2' =>
+    simp
+    rw [←Set.mem_def (a:=s2') (s:=(⋃ (s2 : σ2) (_ : s2 ∈ S2), M2.step s2 x)), Set.mem_iUnion]
+    constructor
+    · rintro ⟨s2, hs2, h2⟩
+      exists s2
+      rw [Set.mem_iUnion]
+      exists hs2
+    · rintro ⟨s2, hs2⟩
+      rw [Set.mem_iUnion] at hs2
+      rcases hs2 with ⟨hs2,h2⟩
+      exists s2
+
+lemma union_start_spec {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    (union M1 M2).start = { s : σ1 ⊕ σ2 | s.casesOn M1.start M2.start } := by rfl
+
+lemma union_acceptsFrom
+ {S1 : Set σ1} {S2 : Set σ2} {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    acceptsFrom (union M1 M2)
+      { s : σ1 ⊕ σ2 | s.casesOn S1 S2 }
+    = M1.acceptsFrom S1 ∪ M2.acceptsFrom S2 := by
+    apply Set.ext
+    intros xs
+    dsimp [acceptsFrom, evalFrom, union, accept]
+    unfold stepSet
+    dsimp [NFA.step]
+    rw [Set.mem_union, Set.mem_setOf, Set.mem_setOf]
+    revert S1 S2
+    induction xs <;> intro S1 S2
+    case nil =>
+      simp
+      simp [Set.mem_def]
+    case cons x xs ih =>
+      simp [List.foldl_cons, List.foldl_cons, List.foldl_cons, ←ih, union_biUnion_spec]
```

Comment:
Please observe the indentation [described in the style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode)

Reply:
My apologies, thanks for catching this!